src/com/facebook/buck/swift/SwiftPrepareForCompile.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2015-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.swift;
+
+import com.facebook.buck.cxx.CxxPrepareForLinkStep;
+import com.facebook.buck.cxx.Linker;
+import com.facebook.buck.io.ProjectFilesystem;
+import com.facebook.buck.model.BuildTargets;
+import com.facebook.buck.rules.AbstractBuildRule;
+import com.facebook.buck.rules.AddToRuleKey;
+import com.facebook.buck.rules.BuildContext;
+import com.facebook.buck.rules.BuildRuleParams;
+import com.facebook.buck.rules.BuildableContext;
+import com.facebook.buck.rules.SourcePath;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.rules.args.Arg;
+import com.facebook.buck.rules.args.FileListableLinkerInputArg;
+import com.facebook.buck.rules.args.SourcePathArg;
+import com.facebook.buck.rules.args.StringArg;
+import com.facebook.buck.step.Step;
+import com.facebook.buck.util.MoreCollectors;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import java.nio.file.Path;
+
+import javax.annotation.Nullable;
+
+class SwiftPrepareForCompile extends AbstractBuildRule {
+
+  @AddToRuleKey
+  private final ImmutableSet<SourcePath> srcs;
+
+  @AddToRuleKey
+  private final Linker linker;
+
+  private final Path fileListPath;
+
+  SwiftPrepareForCompile(
+      BuildRuleParams buildRuleParams,
+      SourcePathResolver resolver,
+      ImmutableSet<SourcePath> srcs,
+      Linker linker) {
+    super(buildRuleParams, resolver);
+    this.srcs = srcs;
+    this.linker = linker;
+    this.fileListPath = BuildTargets.getScratchPath(
+        getProjectFilesystem(), getBuildTarget(), "%s__filelist.txt");
+  }
+
+  private Step createFileListStep() {
+    ImmutableList<Arg> sourceListArgs = FileListableLinkerInputArg.from(
+        SourcePathArg.from(getResolver(), srcs).stream()
```

Comment:
Is this going to work for build target source paths?

If so, then absolute paths would be in the file list, so it shouldn't be cached (see https://github.com/facebook/buck/blob/master/src/com/facebook/buck/rules/SymlinkTree.java#L198-L205 for example).

If not, it should.